### PR TITLE
Add alias command

### PR DESCRIPTION
**Pull request by rachx**

### CODE REVIEW

**Comment by burnflare** [comment]
Assigning to myself to investigate usage of reflection here

**Comment by burnflare** [comment]
@rachx I've pushed updates with reflection working

File: .gitignore [hunk 1]
```diff
@@ -14,3 +14,4 @@ preferences.json
 classes/
 /data/
 /bin/
+/commands.json
```
Comment (rachx) [inline .gitignore]: oops will change after i merged master and updated

Comment (burnflare) [reply .gitignore]: Oh you're changing the file name?

File: src/main/java/seedu/agendum/MainApp.java [hunk 2]
```diff
@@ -152,6 +158,37 @@ protected UserPrefs initPrefs(Config config) {
...
+        
+        //Update commandLibrary file in case it was missing to begin with or there are new/unused fields
+        try {
+            storage.saveCommandLibraryTable(initializedCommandLibraryTable);
+        } catch (IOException e) {
+            logger.warning("Failed to save command library file : " + StringUtil.getDetails(e));
+        }
+        return commandLibrary;
+    }
+
```
Comment (rachx) [inline src/main/java/seedu/agendum/MainApp.java]: Just copying userPrefs and doing everything here

File: src/main/java/seedu/agendum/logic/LogicManager.java [hunk 3]
```diff
@@ -20,17 +23,26 @@
 
     private final Model model;
     private final Parser parser;
+    private CommandLibrary commandLibrary;
 
-    public LogicManager(Model model, Storage storage) {
+    public LogicManager(Model model, Storage storage, CommandLibrary commandLibrary) {
```
Comment (rachx) [inline src/main/java/seedu/agendum/logic/LogicManager.java]: I didn't want to use singleton/static variables when I can ensure that they are all the same commandLibrary this way. Easier for the test cases

File: src/main/java/seedu/agendum/logic/parser/Parser.java [hunk 4]
```diff
@@ -39,13 +39,22 @@
 
     private static final Pattern SCHEDULE_ARGS_FORMAT = Pattern.compile("(?:.+?(?=(?:(?:by|from|to)\\s|$)))+?");
 
+    private static final Pattern ALIAS_ARGS_FORMAT = Pattern.compile(
+            "(?<commandword>[\\p{Alnum}\\p{Punct}]+)\\s+(?<shorthand>[\\p{Alnum}\\p{Punct}]+)");
```
Comment (rachx) [inline src/main/java/seedu/agendum/logic/parser/Parser.java]: make sense to allow a-z, 1-9 and symbols e.g. /?

**Comment by rachx** [comment]
Should be working well and ready for review. Suggest changes to the high-level design too.
Going to clean up more before merging

Autocomplete is a good alternative to this.
Should we allow them to alias an alias? I don't think it is useful.
How about showing all the alias to a command in the help window? Need to make a static method somewhere

File: .gitignore (hunk 1 above)
Comment (rachx) [reply .gitignore]: in the json folder justin created

Comment (burnflare) [reply .gitignore]: Ah cool, just update the `.gitignore` accordingly

File: src/main/java/seedu/agendum/logic/parser/Parser.java (hunk 4 above)
Comment (INCENDE) [reply src/main/java/seedu/agendum/logic/parser/Parser.java]: i think symbols might break the code in someway, if they do think of some really devious methods

**Comment by INCENDE** [comment]
Btw does it support more than one alias? Then i'll have to change the user guide

File: src/main/java/seedu/agendum/logic/parser/Parser.java (hunk 4 above)
Comment (rachx) [reply src/main/java/seedu/agendum/logic/parser/Parser.java]: remove symbols? numbers make sense?

**Comment by rachx** [comment]
Currently it only support
alias <original-command-word> <new-name>
Don't think it will be helpful to let them alias an alias keyword e.g. alias a aa
Yes, it does support multiple alias e.g. m and x for mark

for unalias
unalias <new-name>
Not the original command word

**Comment by INCENDE** [comment]
Ok i'll update user&dev guide to match your description

File: src/main/java/seedu/agendum/logic/commands/UnaliasCommand.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import seedu.agendum.model.Model;
+
+/**
+ * Create an alias for a reserved command keyword
+ */
+public class UnaliasCommand extends Command {
+
+    // COMMAND_WORD, COMMAND_FORMAT, COMMAND_DESCRIPTION are for display in help window
+    public static final String COMMAND_WORD = "unalias";
+    public static String COMMAND_FORMAT = "unalias <your-command>";
```
Comment (burnflare) [inline src/main/java/seedu/agendum/logic/commands/UnaliasCommand.java]: final

File: src/main/java/seedu/agendum/logic/commands/UnaliasCommand.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    public static String COMMAND_DESCRIPTION = "remove your own shorthand command";
```
Comment (burnflare) [inline src/main/java/seedu/agendum/logic/commands/UnaliasCommand.java]: final

File: src/main/java/seedu/agendum/logic/LogicManager.java (hunk 3 above)
Comment (burnflare) [reply src/main/java/seedu/agendum/logic/LogicManager.java]: Cool

File: src/main/java/seedu/agendum/logic/commands/AliasCommand.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import seedu.agendum.model.Model;
+
+/**
+ * Create an alias for a reserved command keyword
+ */
+public class AliasCommand extends Command {
+
+    // COMMAND_WORD, COMMAND_FORMAT, COMMAND_DESCRIPTION are for display in help window
+    public static final String COMMAND_WORD = "alias";
+    public static String COMMAND_FORMAT = "alias <original-command> <your-command>";
```
Comment (burnflare) [inline src/main/java/seedu/agendum/logic/commands/AliasCommand.java]: final

File: src/main/java/seedu/agendum/logic/commands/AliasCommand.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    public static String COMMAND_DESCRIPTION = "specify your own shorthand command";
```
Comment (burnflare) [inline src/main/java/seedu/agendum/logic/commands/AliasCommand.java]: final

**Review by burnflare — APPROVED** [review]
Minor changes requested, after that, can merge.

**Comment by burnflare** [comment]
Fixes #37

**Comment by burnflare** [comment]
Could you address the Codacy `unused local variable` issues before merging? Thanks!

**Comment by rachx** [comment]
don't merge yet. might want to change to singleton after he mentioned doing that for command box history

**Comment by burnflare** [comment]
@rachx What's holding this from being merged?

**Pull request merged**